feat(indexer): Do not reindex objects in checkpoint indexing, fix race conditions for display table

tomxey opened the pull request:
# Description of change

Do not reindex objects in checkpoint indexing, fix race conditions for display table

In general this PR makes sure that it's safe to index old transactions, in a sense that newer data will not be overwritten by older data. (with an exception of concurrent indexing, and indexing transactions for which dependencies are not yet indexed)

This PR is a first step to fix race conditions on `objects` table. To fix them completely https://github.com/iotaledger/iota/issues/7290 and https://github.com/iotaledger/iota/issues/6321 would need to be completed.
This PR fixes race conditions on `display` table completely.

## Links to any relevant issues

fixes https://github.com/iotaledger/iota/issues/7289

## How the change has been tested

`cargo test --profile simulator --package iota-indexer --test rpc-tests --all-features`

- [x] Basic tests (linting, compilation, formatting, unit/integration tests)
- [x] Patch-specific tests (correctness, functionality coverage) - partially tested, race conditions on `display` table are fixed so they are tested, tests for `objects` table can be added later when remaining issues are closed

### Infrastructure QA (only required for crates that are maintained by @iotaledger/infrastructure)

- [ ] Synchronization of the indexer from genesis for a network including migration objects.
- [ ] Restart of indexer synchronization locally without resetting the database.
- [ ] Restart of indexer synchronization on a production-like database.
- [ ] Deployment of services using Docker.
- [ ] Verification of API backward compatibility.

### Release Notes

- [ ] Protocol:
- [ ] Nodes (Validators and Full nodes):
- [ ] Indexer:
- [ ] JSON-RPC:
- [ ] GraphQL:
- [ ] CLI:
- [ ] Rust SDK:
- [ ] REST API:

tomxey committed:
feat(indexer): Do not reindex objects in checkpoint indexing, fix race conditions for display table

sergiupopescu199 commented on crates/iota-indexer/src/handlers/committer.rs:
```diff
@@ -243,3 +263,46 @@ async fn commit_checkpoints<S>(
+                let deleted_objs = tx.effects.all_tombstones();
+                modified_objs
+                    .into_iter()
+                    .map(|o| (o.0.0, o.0.1))
```
nit: if possible, we could add a tuple containing the two values to understand that is `o.0.0` & `o.0.1`

kodemartin replied:
Reverted this logic following #7518

sergiupopescu199 reviewed (approved):
lgtm ✨

kodemartin reviewed (changes requested):
Pending #7518

kodemartin committed:
refactor(indexer): improve on disabling reindexing of optimistic transactions (#7518)

# Description of change

This patch follows #7403 to apply the following improvements:

* Retrieve the transaction digests that touched the objects while
extracting data from `CheckpointData` to avoid lookups afterwards.
* Check if the associated transaction for an object is optimistically
indexed at insertion time, avoiding an additional query on the
application side to filter indexed transaction before the insert. This
eliminates any race conditions that might stem from the two queries as
well.

> [!NOTE]
> A transaction is considered as optimistically index if
`tx_global_order.optimistic_sequence_number > 0`.
>
> We add additional semantics to `optimistic_sequence_number` to avoid
an extra column or different table altogether.

While implementing this, it was challenging to find an ergonomic and
idiomatic way to introduce the changes without breaking the API, or
adding incoherent types in the existing types that would further
complicate proper versioning.

To this end, this patch also introduces the `rolling` module, as a means
to introduce the **next** stable version of the indexer while
maintaining backward compatibility to the existing one. This is
experimental, and the versioning scheme is still to be specified, so
please indulge me by not putting too much focus on its structure which
is subject to change.

## Links to any relevant issues

Part of #7289

## How the change has been tested

Describe the tests that you ran to verify your changes.

- [x] Basic tests (linting, compilation, formatting, unit/integration
tests)
- [ ] Patch-specific tests (correctness, functionality coverage)

### Infrastructure QA (only required for crates that are maintained by
@iotaledger/infrastructure)

Run the `v1.2.3` version of the indexer for a fixed set of checkpoint
from a local network, and then ran the same set of checkpoints against
the version of the indexer introduced with this patch. The resulting
`objects` table were identical.

Other QA tests seemed redundant at this point.

- [ ] Synchronization of the indexer from genesis for a network
including migration objects.
- [ ] Restart of indexer synchronization locally without resetting the
database.
- [ ] Restart of indexer synchronization on a production-like database.
- [ ] Deployment of services using Docker.
- [ ] Verification of API backward compatibility.

kodemartin committed:
refactor(indexer): revert disabling logic from committer
